ppanggolin/annotate/synta.py

```diff
@@ -245,49 +245,65 @@ def launch_infernal(
     return gene_objs
 
 
-def read_fasta(org: Organism, fna_file: Union[TextIOWrapper, list]) -> Dict[str, str]:
-    """Reads a fna file (or stream, or string) and stores it in a dictionary with contigs as key and sequence as value.
+def parse_fasta(
+    fna_file: Union[TextIOWrapper, list]
+) -> Generator[Tuple[str, str], None, None]:
+    """Yields each header and sequence from a FASTA file or stream as a tuple.
 
-    :param org: Organism corresponding to fasta file
-    :param fna_file: Input fasta file with sequences or list of each line as sequence
+    :param fna_file: Input FASTA file or list of lines as sequences.
+    :yield: Tuple with contig header (without '>') and sequence.
+    """
+    header = None
+    sequence = []
+
+    for line in fna_file:
+        line = line.strip()
+        if line.startswith(">"):  # New header
+            if header:  # Yield previous header and sequence
+                yield (header, "".join(sequence))
+            header = line[1:]  # Strip '>'
+            sequence = []
+        else:
+            sequence.append(line)
+
+    if header:  # Yield the final contig
+        yield (header, "".join(sequence))
 
-    :return: Dictionary with contig_name as keys and contig sequence in values
+
+def get_contigs_from_fasta_file(
+    org: Organism, fna_file: Union[TextIOWrapper, list]
+) -> Dict[str, str]:
+    """Processes contigs from a parsed FASTA generator and stores in a dictionary.
+
+    :param org: Organism instance to update with contig info.
+    :param fna_file: Input FASTA file or list of lines as sequences.
+    :return: Dictionary with contig names as keys and sequences as values.
     """
+
     global contig_counter
-    try:
-        contigs = {}
-        contig_seq = ""
-        contig = None
-        for line in fna_file:
-            if line.startswith(">"):
-                if len(contig_seq) >= 1:  # contig filter = 1
-                    contigs[contig.name] = contig_seq.upper()
-                    contig.length = len(contig_seq)
-                contig_seq = ""
-                try:
-                    contig = org.get(line.split()[0][1:])
-                except KeyError:
-                    with contig_counter.get_lock():
-                        contig = Contig(contig_counter.value, line.split()[0][1:])
-                        contig_counter.value += 1
-                    org.add(contig)
-            else:
-                contig_seq += line.strip()
-        if len(contig_seq) >= 1:  # processing the last contig
-            contigs[contig.name] = contig_seq.upper()
-            contig.length = len(contig_seq)
-
-    except AttributeError as e:
-        raise AttributeError(
```

Comment:
In case we provide a file that is not a fasta while saying that it is a fasta, is there an error raised before that deals with it ?

Reply:
Good point. I added some check and pytest on this matter.